blog/_posts/2016-02-01-iteration.md

```diff
@@ -0,0 +1,420 @@
+relationship between these iterators and
+[`Base.Cartesian`](http://docs.julialang.org/en/stable/devdocs/cartesian/),
+which is a completely different (and much more painful) approach to
+solving the same problem.  There are still a few occasions where
+`Base.Cartesian` is necessary, but for many problems these new
+capabilities represent a vastly simplified approach.
+
+Let's introduce these iterators with an extension of an example taken
+from the
+[manual](http://docs.julialang.org/en/stable/manual/arrays/#iteration).
+
+## `eachindex`, `CartesianIndex`, and `CartesianRange`
+
+You may already know that, in julia 0.4, there are two recommended
+ways to iterate over the elements in an `AbstractArray`: if you don't
+need an index associated with each element, then you can use
+
+```jl
+for a in A    # A is an AbstractArray
+    # Code that does something with the element a
+end
+```
+
+If instead you also need the index, then use
+
+```jl
+for i in eachindex(A)
+    # Code that does something with i and/or A[i]
+end
+```
+
+In some cases, the first line of this loop expands to `for i =
+1:length(A)`, and `i` is just an integer.  However, in other cases,
+this will expand to the equivalent of
+
+```jl
+for i in CartesianRange(size(A))
+    # i is now a CartesianIndex
+    # Code that does something with i and/or A[i]
+end
+```
+
+Let's see what these objects are:
+
+```jl
+A = rand(3,2)
+
+julia> for i in CartesianRange(size(A))
+           @show i
+       end
+i = CartesianIndex{2}((1,1))
+i = CartesianIndex{2}((2,1))
+i = CartesianIndex{2}((3,1))
+i = CartesianIndex{2}((1,2))
+i = CartesianIndex{2}((2,2))
+i = CartesianIndex{2}((3,2))
+```
+
+A `CartesianIndex{N}` represents an `N`-dimensional index.
+`CartesianIndex`es are based on tuples, and indeed you can access the
+underlying tuple with `i.I`.  However, they also support certain
+arithmetic operations, treating their contents like a fixed-size
+`Vector{Int}`.  Since the length is fixed, julia/LLVM can generate
+very efficient code (without introducing loops) for operations with
+N-dimensional `CartesianIndex`es.
+
+A `CartesianRange` is just a pair of `CartesianIndex`es, encoding the
+start and stop values along each dimension, respectively:
+
+```jl
+julia> CartesianRange(size(A))
+CartesianRange{CartesianIndex{2}}(CartesianIndex{2}((1,1)),CartesianIndex{2}((3,2)))
+```
+
+You can construct these manually: for example,
+
+```jl
+julia> CartesianRange(CartesianIndex((-7,0)), CartesianIndex((7,15)))
+CartesianRange{CartesianIndex{2}}(CartesianIndex{2}((-7,0)),CartesianIndex{2}((7,15)))
+```
+
+constructs a range that will loop over `-7:7` along the first
+dimension and `0:15` along the second.
+
+One reason that `eachindex` is recommended over `for i = 1:length(A)`
+is that some `AbstractArray`s cannot be indexed efficiently with a
+linear index; in contrast, a much wider class of objects can be
+efficiently indexed with a multidimensional iterator.  (SubArrays are,
```

Comment:
GitHub didn't auto-close this from the commit, but I addressed this too, by linking to [devdocs](http://docs.julialang.org/en/stable/devdocs/subarrays/#indexing-cartesian-vs-linear-indexing).